content/blog/2019-10-23-sccp.md

```diff
@@ -0,0 +1,263 @@
+manner that would accurately reflect its utility. Upon running the optimization
+on several test cases, I have observed that, for programs that are written
+directly by humans in Bril, the optimization tends to exhibit one of two
+behaviors: either the program is effectively unchanged, or it is completely
+evaluated leaving only assignments of constants to variables, and print
+statements with those variables. By running these test cases, however, I have
+been able to ensure that the optimization does not change the behavior of
+well-typed Bril programs.
+
+For example, of the test cases provided in the [bril-benchmark][] repository,
+two (`factorial` and `fibonacci`) are completely unchanged by the optimization
+modulo the order of basic blocks, and the other two (`matrix_mul` and
+`poly_mul`) are completely evaluated to just constant assignments and print
+statements. The factor that seems to separate these two classes of programs is
+the existance or lack of loops. This is because, if there are no loops, then
+every assignment statement occurs no more than once. As such, because vanilla
+Bril has no channels through which data can enter a function from the outside,
+every variable's value can be determined through the conditional constant
+propagation analysis. When loops are involved, variables' values change between
+iterations, and as such the analysis is unable to determine a constant value for
+the variables.
+
+I wrote a few programs in TypeScript, and used the <code>[ts2bril][]</code>
+compiler to convert them to Bril. These converted programs are qualitatively
+different from the handwritten programs, as the compiler inserts a lot of
+short-lived variables in order to translate contructs from the TypeScript
+language. Due to the high number of redundant variables, I believe a more
+effective optimization for working on programs outputted by this compiler would
+be [copy propagation][]. For a TypeScript program that prints the first 20
+fibonacci numbers, the constant propagation optimization had the following
+effect:
+
+<div align="center">
+
+| Instruction Count | Unoptimized | Optimized | Change |
+|-------------------|------------:|----------:|-------:|
+| Static            |          26 |        23 | –11.5% |
+| Dynamic           |         410 |       388 |  –5.4% |
+
+</div>
+
+For this program, the optimization removed only three instructions, one of which
+was in the loop body (an extraneous assignment to a variable that was never
+used). In general, it is difficult to judge the effectiveness of this
+optimization without the infrastructure of other optimizations and compilers to
+pair it with. In the future, if someone were to implement, say, a function
+inlining optimization, and were to extend this optimization to support function
+calls, I would be interested to see the effect it would have.
+
+--------------------------------------------------------------------------------
```

Comment:
FWIW, we added some style to footnotes in the CSS for the blog posts such that a visual separator may not be necessary.